File: Problems/129. Sum Root to Leaf Numbers/sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

//Solution 1: 0ms 40.95 MB
class Solution {
  public int sumNumbers(TreeNode root) {
    return calculatePathSum(root, 0);
  }

  public int calculatePathSum(TreeNode node, int currentSum) {
    if (node == null)
      return 0;

    int newPathSum = currentSum * 10 + node.val;

    if (node.left == null && node.right == null)
      return newPathSum;

    return calculatePathSum(node.left, newPathSum)
        + calculatePathSum(node.right, newPathSum);
  }
}

File: Problems/129. Sum Root to Leaf Numbers/sum-root-to-leaf-numbers.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

//Solution 1: 35ms 52.00 MB
function sumNumbers(root: TreeNode | null): number {
  return calculatePathSum(root, 0);
}

function calculatePathSum(node: TreeNode | null, currentSum: number): number {
  if (!node) return 0;

  const newPathSum = currentSum * 10 + node.val;

  if (!node.left && !node.right) {
    return newPathSum;
  }

  return (
    calculatePathSum(node.left, newPathSum) +
    calculatePathSum(node.right, newPathSum)
  );
}
